### PR TITLE
[FLINK-17451][sql-parser][table-planner-blink][hive] Implement view D…

### PR DESCRIPTION
**Pull request by lirui-apache**
…DLs for Hive dialect

## What is the purpose of the change

Implement view DDLs for Hive dialect.

## Brief change log

  - Implement view DDL syntax for hive dialect.
  - Add SqlNodes for ALTER VIEW statements.
  - Add AlterOperation for ALTER VIEW AS statement.
  - Update `SqlToOperationConverter` and `TableEnvironmentImpl` to support alter view operations.
  - Add test cases.

## Verifying this change

Existing and added tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? docs

### CODE REVIEW

**Comment by lirui-apache** [comment]
@JingsongLi @danny0405 Please have a look. Thanks.

File: flink-table/flink-sql-parser-hive/src/main/codegen/includes/parserImpls.ftl [hunk 1]
```diff
@@ -958,3 +958,90 @@ void TableApiIdentifierSegment(List<String> names, List<SqlParserPos> positions)
         }
     }
 }
+
+/*  Parses a create view or replace existing view statement.
+ *  CREATE [OR REPLACE] VIEW view_name [ (field1, field2 ...) ] AS select_statement
```
Comment (KurtYoung) [inline flink-table/flink-sql-parser-hive/src/main/codegen/includes/parserImpls.ftl:963]: Provide an example in hive dialect

Comment (KurtYoung) [reply flink-table/flink-sql-parser-hive/src/main/codegen/includes/parserImpls.ftl:963]: BTW, I think it's better to copy and paste some hive's DDL LanguageManual to each supported statement kind.

File: flink-table/flink-sql-parser-hive/src/main/java/org/apache/flink/sql/parser/hive/ddl/SqlCreateHiveView.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+import org.apache.calcite.sql.SqlCharStringLiteral;
+import org.apache.calcite.sql.SqlIdentifier;
+import org.apache.calcite.sql.SqlNode;
+import org.apache.calcite.sql.SqlNodeList;
+import org.apache.calcite.sql.SqlWriter;
+import org.apache.calcite.sql.parser.SqlParserPos;
+
+/**
+ * CREATE View DDL for Hive dialect.
```
Comment (KurtYoung) [inline flink-table/flink-sql-parser-hive/src/main/java/org/apache/flink/sql/parser/hive/ddl/SqlCreateHiveView.java:32]: put hive's Language Manual here, easier for us to check whether this class make sense.

File: flink-table/flink-sql-parser-hive/src/main/codegen/includes/parserImpls.ftl (hunk 1 above)
Comment (lirui-apache) [reply flink-table/flink-sql-parser-hive/src/main/codegen/includes/parserImpls.ftl]: I have added Hive syntax in the comments. Note that some features are not yet supported, e.g. column comments for views.

File: flink-table/flink-sql-parser-hive/src/test/java/org/apache/flink/sql/parser/hive/FlinkHiveSqlParserImplTest.java [hunk 3]
```diff
@@ -221,4 +221,49 @@ public void testInsert() {
...
+				.ok("CREATE VIEW `V1`\n" +
+						"COMMENT 'v1 comment'\n" +
+						"TBLPROPERTIES (\n" +
+						"  'k1' = 'v1',\n" +
+						"  'k2' = 'v2'\n" +
+						")\n" +
+						"AS\n" +
+						"SELECT *\n" +
+						"FROM `TBL`");
+		// TODO: support column comments
```
Comment (KurtYoung) [inline flink-table/flink-sql-parser-hive/src/test/java/org/apache/flink/sql/parser/hive/FlinkHiveSqlParserImplTest.java:247]: will column comment supported in 1.11?

Comment (lirui-apache) [reply flink-table/flink-sql-parser-hive/src/test/java/org/apache/flink/sql/parser/hive/FlinkHiveSqlParserImplTest.java:247]: I'm afraid not. Actually table's column comment is not supported neither. The comments get discarded when converted to `TableSchema`.

**Comment by lirui-apache** [comment]
The `test_ci tests` failure shouldn't be related because it has passed in my personal build:
https://dev.azure.com/lirui-apache/flink/_build/results?buildId=61&view=results

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/ddl/AlterViewAsOperation.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+package org.apache.flink.table.operations.ddl;
+
+import org.apache.flink.table.catalog.CatalogView;
+import org.apache.flink.table.catalog.ObjectIdentifier;
+
+/**
+ * Operation to describe an ALTER VIEW ... AS ... statement.
+ */
+public class AlterViewAsOperation extends AlterTableOperation {
```
Comment (KurtYoung) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/ddl/AlterViewAsOperation.java:27]: create a base class `AlterViewOperation` and extends from that, `AlterViewOperation` should be the same level of `AlterTableOperation `

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/ddl/AlterTablePropertiesOperation.java [hunk 5]
```diff
@@ -26,27 +27,28 @@
 
 
 /**
- * Operation to describe a ALTER TABLE .. SET .. statement.
+ * Operation to describe a ALTER TABLE/VIEW .. SET .. statement.
  */
 public class AlterTablePropertiesOperation extends AlterTableOperation {
```
Comment (KurtYoung) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/ddl/AlterTablePropertiesOperation.java:32]: Create a dedicated `AlterViewPropertiesOperation` and extends from `AlterViewOperation`

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/operations/SqlToOperationConverter.java [hunk 6]
```diff
@@ -227,86 +234,134 @@ private Operation convertDropTable(SqlDropTable sqlDropTable) {
...
+		}
+		CatalogBaseTable baseTable = optionalCatalogTable.get().getTable();
+		if (baseTable instanceof CatalogTable) {
+			throw new ValidationException("ALTER VIEW for a table is not allowed");
+		}
+		if (alterView instanceof SqlAlterViewRename) {
+			UnresolvedIdentifier newUnresolvedIdentifier =
+					UnresolvedIdentifier.of(((SqlAlterViewRename) alterView).fullNewViewName());
+			ObjectIdentifier newTableIdentifier = catalogManager.qualifyIdentifier(newUnresolvedIdentifier);
+			return new AlterTableRenameOperation(viewIdentifier, newTableIdentifier);
```
Comment (KurtYoung) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/operations/SqlToOperationConverter.java:256]: create a new `AlterViewRenameOperation`

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/operations/SqlToOperationConverter.java [hunk 7]
```diff
@@ -227,86 +234,134 @@ private Operation convertDropTable(SqlDropTable sqlDropTable) {
...
+			CatalogView oldView = (CatalogView) baseTable;
+			Map<String, String> newProperties = new HashMap<>(oldView.getOptions());
+			newProperties.putAll(extractProperties(alterViewProperties.getPropertyList()));
+			CatalogView newView = new CatalogViewImpl(
+					oldView.getOriginalQuery(),
+					oldView.getExpandedQuery(),
+					oldView.getSchema(),
+					newProperties,
+					oldView.getComment());
+			return new AlterTablePropertiesOperation(viewIdentifier, newView);
```
Comment (KurtYoung) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/operations/SqlToOperationConverter.java:268]: don't reuse `AlterTablePropertiesOperation `

**Pull request closed without merging**